res/res_pjsip_device_features_body_generator.c

```diff
@@ -0,0 +1,325 @@
+
+	/*
+	 * The phone is expecting something like this:
+	 *
+	 * <?xml version="1.0" encoding="ISO-8859-1"?>
+	 * <ForwardingEvent xmlns="http://www.ecmainternational.org/standards/ecma-323/csta/ed3">
+	 *    <device>5559430902</device>
+	 *    <forwardingType>forwardImmediate</forwardingType>
+	 *    <forwardStatus>false</forwardStatus>
+	 * </ForwardingEvent>
+	 *
+	 * <?xml version="1.0" encoding="ISO-8859-1"?>
+	 * <ForwardingEvent xmlns="http://www.ecmainternational.org/standards/ecma-323/csta/ed3">
+	 *    <device>5559430902</device>
+	 *    <forwardingType>forwardNoAns</forwardingType>
+	 *    <forwardStatus>false</forwardStatus>
+	 *    <forwardTo></forwardTo>
+	 *    <ringCount></ringCount>
+	 * </ForwardingEvent>
+	 *
+	 * ringcount, if present, can range from 1-100
+	 * Currently this isn't something we send to the phone, since the forward time is handled server side anyways.
+	 */
+
+	tmpnode = ast_xml_new_node("device");
+	if (!tmpnode) {
+		goto cleanup;
+	}
+	/* The Broadworks spec says device is mandatory, but the actual value is used by neither server nor client. So if we don't have one, make something up. */
+	ast_xml_set_text(tmpnode, S_OR(sync_data->deviceid, "123456"));
+	if (!sync_data->deviceid[0]) {
+		ast_debug(2, "This was the first NOTIFY with body data for this endpoint\n");
+	}
+	ast_xml_add_child(root, tmpnode);
+
+	tmpnode = ast_xml_new_node("forwardingType");
+	if (!tmpnode) {
+		goto cleanup;
+	}
+	ast_xml_set_text(tmpnode, fwd_str);
+	ast_xml_add_child(root, tmpnode);
+
+	active = !ast_strlen_zero(forward_to) ? 1 : 0;
+
+	tmpnode = ast_xml_new_node("forwardStatus");
+	if (!tmpnode) {
+		goto cleanup;
+	}
+	ast_xml_set_text(tmpnode, active ? "true" : "false");
+	ast_xml_add_child(root, tmpnode);
+
+	if (active) { /* Forwarding is active. */
+		tmpnode = ast_xml_new_node("forwardTo");
+		if (!tmpnode) {
+			goto cleanup;
+		}
+		ast_xml_set_text(tmpnode, forward_to);
+		ast_xml_add_child(root, tmpnode);
+	} /* else, forwarding is not active. No additional info needed. */
+
+	/* Finalize. */
+	ast_xml_doc_dump_memory(doc, &buf, &len);
+	ast_xml_close(doc);
+	if (len <= 0) {
+		ast_log(LOG_WARNING, "XML document has length %d?\n", len);
+	}
+	if (buf) {
+		ast_str_append(bodytext, 0, "%s", buf);
+		ast_xml_free_text(buf);
+	}
+	return 0;
+cleanup:
+	ast_xml_close(doc);
+	ast_log(LOG_ERROR, "Could not create new XML root node\n");
+	return -1;
+}
+
+static int features_generate_body_content(void *body, void *data)
+{
+	char *buf;
+	int len;
+	struct ast_str **bodytext = body;
+	struct ast_xml_doc *doc;
+	struct ast_xml_node *root, *tmpnode;
+	struct ast_sip_device_feature_sync_data *sync_data = data;
+
+	int updates_made = 0;
+
+	/* This callback is called for *all* NOTIFYs, so we should only add XML to the body if actually necessary. */
+	ast_debug(2, "Generating body content for %s/%s\n", FEATURE_TYPE, FEATURE_SUBTYPE);
```

Comment:
This needs more information such as the endpoint, and so on for other log messages, and meaningless messages for users too